cmd/contour/serve.go

```diff
@@ -374,50 +369,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-		} else {
-			informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.ServiceAPIResources()...)
+		for _, r := range k8s.ServiceAPIResources() {
+			if !clients.ResourcesExist(r) {
+				log.WithField("resource", r).Warn("resource type not present on API server")
+				continue
+			}
+			inf, err := clients.InformerForResource(r)
+			if err != nil {
+				log.WithError(err).WithField("resource", r).Fatal("failed to create informer")
+			}
+
+			inf.AddEventHandler(dynamicHandler)
 		}
 	}
 
-	// TODO(youngnick): Move this logic out to internal/k8s/informers.go somehow.
-	// Add informers for each root namespace
-	for _, factory := range namespacedInformerFactories {
-		informerSyncList.InformOnResources(factory, dynamicHandler, k8s.SecretsResources()...)
-	}
+	// Inform on secrets, filtering by root namespaces.
+	for _, r := range k8s.SecretsResources() {
+		inf, err := clients.InformerForResource(r)
+		if err != nil {
+			log.WithError(err).WithField("resource", r).Fatal("failed to create informer")
+		}
 
-	// If root namespaces are not defined, then add the informer for all namespaces
-	if len(namespacedInformerFactories) == 0 {
-		informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.SecretsResources()...)
+		if len(informerNamespaces) == 0 {
+			inf.AddEventHandler(dynamicHandler)
+		} else {
+			inf.AddEventHandler(k8s.NewNamespaceFilter(informerNamespaces, dynamicHandler))
+		}
 	}
 
-	informerSyncList.InformOnResources(clusterInformerFactory,
-		&k8s.DynamicClientHandler{
+	// Inform on endpoints.
+	for _, r := range k8s.EndpointsResources() {
+		inf, err := clients.InformerForResource(r)
+		if err != nil {
+			log.WithError(err).WithField("resource", r).Fatal("failed to create informer")
+		}
+
+		inf.AddEventHandler(&k8s.DynamicClientHandler{
 			Next: &contour.EventRecorder{
 				Next:    endpointHandler,
 				Counter: contourMetrics.EventHandlerOperations,
 			},
 			Converter: converter,
 			Logger:    log.WithField("context", "endpointstranslator"),
-		}, k8s.EndpointsResources()...)
+		})
+	}
 
 	// Set up workgroup runner and register informers.
 	var g workgroup.Group
-	g.Add(startInformer(clusterInformerFactory, log.WithField("context", "contourinformers")))
 
-	for ns, factory := range namespacedInformerFactories {
-		g.Add(startInformer(factory, log.WithField("context", "corenamespacedinformers").WithField("namespace", ns)))
-	}
+	// Register a task to start all the informers.
+	g.Add(func(stop <-chan struct{}) error {
+		log := log.WithField("context", "informers")
+
+		log.Info("starting informers")
+		defer log.Println("stopped informers")
+
+		if err := clients.StartInformers(stop); err != nil {
+			log.WithError(err).Error("failed to start informers")
+		}
+
+		<-stop
+		return nil
+	})
 
 	// Register our event handler with the workgroup.
 	g.Add(eventHandler.Start())
```

Comment:
I *think* that we should probably be waiting for the informer caches to sync in here before starting to process events to avoid any race conditions with the caches populating, but I'd have to experiment a bit to confirm if needed/not problematic.

Reply:
Can you elaborate why you think this is needed? This change keeps the current semantics of not starting the xDS server until the informer cache is present. That's not awesome, but it's not different than what we had.

Reply:
https://github.com/projectcontour/contour/issues/1280

Reply:
Typically in Kubernetes controllers, you would wait for the informer caches to sync (via WaitForCacheSync) before starting to process events from those informers. I think the reason for this is that if you don't, then you may get not-found errors if you try to look up other items from the informer cache as part of your event handler logic.

One scenario I'm thinking of is in the load balancer address update code, which handles an event for the load balancer service, then tries to look up all ingresses/proxies from the cache so it can write addresses to them. if the ingress/informer caches haven't fully synced yet, then we could get an incomplete list back and not update all of them as expected. 

Practically it may or may not be an issue for us, I'm not sure. If we're not seeing issues, then we can defer/ignore.

Reply:
> One scenario I'm thinking of is in the load balancer address update code, which handles an event for the load balancer service, then tries to look up all ingresses/proxies from the cache so it can write addresses to them. if the ingress/informer caches haven't fully synced yet, then we could get an incomplete list back and not update all of them as expected.

I don't think this is a problem. The List is for updating objects that won't get passed to `OnAdd`. If the cache isn't synced when the List happens, then the `OnAdd` would work.

Reply:
The current code does wait for the caches to be synced, but doesn't wait for the first dag rebuild to happen before starting the xds server. Need to make sure that behavior is still kept. Contour needs to wait for the caches to be synced before starting things up. The 2nd bit is still an open issue that needs solved.

Reply:
IIUC, we all agree that this PR doesn't change the current Contour semantics (even though those semantics leave something to be desired)? If so, then we can resolve this thread.

Reply:
filed #3020